src/coreclr/src/vm/jithelpers.cpp

```diff
@@ -5096,6 +5096,426 @@ HCIMPL0(void, JIT_DebugLogLoopCloning)
 }
 HCIMPLEND
 
+#ifdef FEATURE_ON_STACK_REPLACEMENT
+
+// Runtime state for a patchpoint.
+//
+// Patchpoint identity is currently the return address
+// of the helper call in the jitted code.
+//
+// Todo: handle cases where IP addresses get recycled
+struct PerPatchpointInfo
+{
+    PerPatchpointInfo() : 
+        m_osrMethodCode(0),
+        m_patchpointCount(0),
+        m_flags(0),
+        m_patchpointId(0)
+    {
+    }
+
+    enum 
+    {
+        patchpoint_triggered = 0x1,
+        patchpoint_invalid = 0x2
+    };
+
+    PCODE m_osrMethodCode;
+    LONG m_patchpointCount;
+    LONG m_flags;
+
+    int m_patchpointId;
+};
+
+typedef EEPtrHashTable JitPatchpointTable;
+JitPatchpointTable *g_pJitPatchpointTable = NULL;
+CrstStatic g_pJitPatchpointCrst;
+static int s_patchpointId = 0;
+static bool s_warnedLimitedRange = false;
+
+// Helper method to jit the OSR version of a method.
+//
+// Returns the address of the jitted code.
+// Returns NULL if osr method can't be created.
+static PCODE JitPatchpointWorker(void* ip, int ilOffset)
+{
+    PCODE osrVariant = NULL;
+
+    GCX_PREEMP();
+
+    // Find the method desc for this bit of code
+    EECodeInfo codeInfo((PCODE)ip);
+    MethodDesc* pMD = codeInfo.GetMethodDesc();
+
+    // Find the il method version corresponding to this version of the method
+    // and set up a new native code version for it.
+    ReJITID rejitId = ReJitManager::GetReJitId(pMD, codeInfo.GetStartAddress());
+    CodeVersionManager* codeVersionManager = pMD->GetCodeVersionManager();
+    NativeCodeVersion osrNativeCodeVersion;
+    {
+        // Request a new native version that is optimized.
+        CodeVersionManager::TableLockHolder lock(codeVersionManager);
+        ILCodeVersion ilCodeVersion = codeVersionManager->GetILCodeVersion(pMD, rejitId);
+        HRESULT hr = ilCodeVersion.AddNativeCodeVersion(pMD, NativeCodeVersion::OptimizationTier1, &osrNativeCodeVersion);
```

Comment:
Due to the check mentioned by @noahfalk and since the OSR code appears to be specific to each patchpoint, it may prevent a later normal transition to full tier 1 method code. The OSR code could have a different tier identifier but for now it may work to classify it as `OptimizationTier0` instead.

Adding the code version to this collection is also a bit odd because it's not a version of code for the method, and it looks like there are already separate data structures tracking OSR code versions. I don't think it hurts much, for example the debugger would show OSR code versions as well. But I'm curious if a code version needs to be added at all, is it used somehow and if so could those cases use the other data structures instead?

Reply:
> is also a bit odd because it's not a version of code for the method

I suggest that we do think of OSR compilations as 'a version of code for the method'. It is a version that has a different entrypoint and ABI than we are used to, but for the majority of code interacting with CodeVersionManager that isn't an important distinction.

> for example the debugger would show OSR code versions as well

That is a good thing : ) The debugger, profiler, and ETW should all be able to enumerate OSR code bodies as potential code regions that resolve to the same method.

> I'm curious if a code version needs to be added at al

I believe it is an apt representation:
a) It is the identifier used by PrepareCode that identifies the code that needs to be created.
b) It allows the debugger, profiler, and ETW to enumerate the code regions

> if so could those cases use the other data structures instead?

I'm sure it is possible, but the code complexity added to the patchpoint table, in PrepareCode, and in all the code version enumerators goes up significantly when we don't use NativeCodeVersion as a common abstraction. The complexity added to NativeCodeVersion to let it represent an OSR version appears far lower.

Reply:
Am going to stick with a new native code version for now. This is no longer described as Tier1 -- it's now Tier1OSR, so should not confuse tiering...